akka-stream/src/main/scala/akka/stream/scaladsl/One2OneBidiFlow.scala

```diff
@@ -77,10 +75,7 @@ object One2OneBidiFlow {
 
       setHandler(outOut, new OutHandler {
         override def onPull(): Unit = pull(outIn)
-        override def onDownstreamFinish(): Unit = {
-          cancel(outIn)
-          cancel(inIn) // short-cut to speed up cleanup of upstream
-        }
+        override def onDownstreamFinish(): Unit = cancel(outIn)
```

Comment:
point 4) above is not true, it will signal UnexpectedOutputException in this case

`GraphStageLogic.complete(...)` should get the same treatment of Emitting states that `completeStage()` has, I suspect that this may be the cause of some problems you see

`pullsSuppressed` can only ever be 0 or 1, i.e. it should be a boolean